pkg/pillar/types/watcher.go

```diff
@@ -0,0 +1,199 @@
+	RedZone UsageZone = 3
+)
+
+// MemoryNotification :
+type MemoryNotification struct {
+	Total uint64 // Total memory in Bytes
+	Used  uint64 // Used memory in Bytes
+	Zone  UsageZone
+
+	PrevUsage uint64   // Previous (last) memory usage in Bytes
+	LastFive  []uint64 // Last 5 Usage percentage values
+
+	UsageSlab uint64
+	PrevSlab  uint64
+}
+
+func zonetoString(zone UsageZone) string {
+	switch zone {
+	case GreenZone:
+		return "GREEN"
+	case YellowZone:
+		return "YELLOW"
+	case OrangeZone:
+		return "ORANGE"
+	case RedZone:
+		return "RED"
+	}
+	return "UNKNOWN"
+}
+
+// LogCreate :
+func (mem MemoryNotification) LogCreate(logBase *base.LogObject) {
+	logObject := base.NewLogObject(logBase, base.MemoryNotificationType, "MemoryNotification",
+		uuid.UUID{}, mem.LogKey())
+	if logObject == nil {
+		return
+	}
+	loggable := logObject.CloneAndAddField("Total", mem.Total).
+		CloneAndAddField("Used", mem.Used).
+		CloneAndAddField("Zone", zonetoString(mem.Zone)).
+		CloneAndAddField("PrevUsage", mem.PrevUsage).
+		CloneAndAddField("LastFive", mem.LastFive).
+		CloneAndAddField("UsageSlab", mem.UsageSlab).
+		CloneAndAddField("PrevSlab", mem.PrevSlab)
+	switch mem.Zone {
+	case GreenZone, YellowZone:
+		loggable.Functionf("MemoryNotification create")
+	case OrangeZone:
+		loggable.Warnf("MemoryNotification create")
+	case RedZone:
+		loggable.Errorf("MemoryNotification create")
+	default:
+		loggable.Noticef("MemoryNotification create")
+	}
+}
+
+// LogModify :
+func (mem MemoryNotification) LogModify(logBase *base.LogObject, old interface{}) {
+	logObject := base.NewLogObject(logBase, base.MemoryNotificationType, "MemoryNotification",
+		uuid.UUID{}, mem.LogKey())
+	if logObject == nil {
+		return
+	}
+	oldMem, ok := old.(MemoryNotification)
+	if !ok {
+		return
+	}
+	if mem.UsageSlab == oldMem.UsageSlab && mem.Zone == oldMem.Zone {
+		return
+	}
+	loggable := logObject.CloneAndAddField("Total", mem.Total).
+		CloneAndAddField("Used", mem.Used).
+		CloneAndAddField("Zone", zonetoString(mem.Zone)).
+		CloneAndAddField("PrevUsage", mem.PrevUsage).
+		CloneAndAddField("LastFive", mem.LastFive).
+		CloneAndAddField("UsageSlab", mem.UsageSlab).
+		CloneAndAddField("PrevSlab", mem.PrevSlab)
+	switch mem.Zone {
+	case GreenZone, YellowZone:
+		loggable.Functionf("MemoryNotification create")
+	case OrangeZone:
+		loggable.Warnf("MemoryNotification create")
+	case RedZone:
+		loggable.Errorf("MemoryNotification create")
+	default:
+		loggable.Noticef("MemoryNotification create")
+	}
+}
+
+// LogDelete :
+func (mem MemoryNotification) LogDelete(logBase *base.LogObject) {
+}
+
+// LogKey :
+func (mem MemoryNotification) LogKey() string {
+	return "MemoryNotification"
+}
+
+// DiskNotification :
+type DiskNotification struct {
+	Total uint64 // Total Disk space in Bytes
+	Used  uint64 // Used Disk space in Bytes
+	Zone  UsageZone
+
+	PrevUsage uint64   // Previous (last) disk usage in Bytes
+	LastFive  []uint64 // Last 5 usage percentage values
```

Comment:
I don't see code which uses lastFive for anything but logging. Is that the intent?

Reply:
Yes, it is currently information purpose only. Shows the progression of usage changes when there are slab or zone changes.